internal/safebuffer/safereader/safereader.go

```diff
@@ -0,0 +1,160 @@
+package safereader
+
+import (
+	"bytes"
+	"io"
+	"log/slog"
+	"sync"
+)
+
+// Reader is a goroutine safe io.ReadCloser that makes initial data available to
+// callers as well as new data that arrives later via channel.
+type Reader struct {
+	mu        sync.Mutex
+	buf       *bytes.Buffer
+	close     func()
+	closed    <-chan struct{}
+	dataCh    <-chan []byte
+	dataAvail chan struct{}
+	closeOnce sync.Once
+}
+
+var _ io.ReadCloser = (*Reader)(nil)
+
+// jobIsRunning doesn't have a race because Buffer.NewReader calls it while
+// holding a lock that prevents new data, or the close of jobDone, from being
+// processed.
+func jobIsRunning(jobDone <-chan struct{}) bool {
+	select {
+	case <-jobDone:
+		return false
+	default:
+		return true
+	}
+}
+
+type Channels struct {
+	Data   chan []byte
+	Closed <-chan struct{}
+}
+
+// New returns a new Reader that keeps a copy of data and will append any data
+// that arrives on the returned Data channel. If the job is no longer running
+// when created, it is treated as a closed reader and no new data can be added.
+// Calls to Close() in this case are ignored.
+func New(data []byte, jobDone <-chan struct{}) (*Reader, *Channels) {
+	closed := make(chan struct{})
+	closeFn := func() { close(closed) }
+
+	buf := make([]byte, len(data))
+	copy(buf, data)
+
+	r := Reader{
+		close:  closeFn,
+		closed: closed,
+		buf:    bytes.NewBuffer(buf),
+	}
+
+	var channels *Channels
+
+	if jobIsRunning(jobDone) {
+		dataCh := make(chan []byte)
+		channels = &Channels{
+			Data:   dataCh,
+			Closed: closed,
+		}
+		r.dataCh = dataCh
+		r.dataAvail = make(chan struct{})
+
+		go r.receiver()
```

Comment:
is it possible to live without spinning extra go-routines?

Reply:
With the current design, it's one goroutine per reader, so it's not awful, but it is required to prevent blocking/slow writes. I suppose I could have used buffered channels to fix _most_ blocking issues, but it's still not a guarantee that they will always be avoided. Generally, I view buffered channels as a bit of a smell (except in some cases where the buffer is 1 so that writes don't block forever if a reader disappears) that the code could probably be improved. Besides, the cost of launching a single goroutine is very small, it needs to be managed properly, but I don't see it as a major concern.